examples/systemcalls/systemcalls.c

```diff
@@ -1,97 +1,128 @@
 {
     va_list args;
     va_start(args, count);
-    char * command[count+1];
+    char *command[count + 1];
     int i;
-    for(i=0; i<count; i++)
+    for (i = 0; i < count; i++)
     {
         command[i] = va_arg(args, char *);
     }
     command[count] = NULL;
-    // this line is to avoid a compile warning before your implementation is complete
-    // and may be removed
-    command[count] = command[count];
-
-/*
- * TODO:
- *   Execute a system command by calling fork, execv(),
- *   and wait instead of system (see LSP page 161).
- *   Use the command[0] as the full path to the command to execute
- *   (first argument to execv), and use the remaining arguments
- *   as second argument to the execv() command.
- *
-*/
 
     va_end(args);
 
-    return true;
+    pid_t pid;
+    pid = fork();
+
+    if (pid == -1)
+    {
+        return false;
+    }
+    else if (pid == 0)
+    {
+        execv(command[0], command);
+        exit(99);
+    }
+    else
+    {
+        int status;
+        waitpid(pid, &status, 0);
+        if (WIFEXITED(status) && (WEXITSTATUS(status) == 99) | (WEXITSTATUS(status) == 1))
+            return false;
+
+        else
+            return true;
+    }
 }
 
 /**
-* @param outputfile - The full path to the file to write with command output.
-*   This file will be closed at completion of the function call.
-* All other parameters, see do_exec above
-*/
+ * @param outputfile - The full path to the file to write with command output.
+ *   This file will be closed at completion of the function call.
+ * All other parameters, see do_exec above
+ */
 bool do_exec_redirect(const char *outputfile, int count, ...)
 {
     va_list args;
     va_start(args, count);
-    char * command[count+1];
+    char *command[count + 1];
     int i;
-    for(i=0; i<count; i++)
+    for (i = 0; i < count; i++)
     {
         command[i] = va_arg(args, char *);
     }
     command[count] = NULL;
-    // this line is to avoid a compile warning before your implementation is complete
-    // and may be removed
-    command[count] = command[count];
 
+    pid_t pid;
+    int fd = open(outputfile, O_TRUNC | O_CREAT | O_WRONLY, 0644);
+
+    pid = fork();
 
-/*
- * TODO
- *   Call execv, but first using https://stackoverflow.com/a/13784315/1446624 as a refernce,
- *   redirect standard out to a file specified by outputfile.
- *   The rest of the behaviour is same as do_exec()
- *
-*/
+    if (fd < 0)
+        return false;
+
+    if (pid == -1)
+    {
+        return false;
+    }
+    else if (pid == 0)
+    {
+        if (dup2(fd, 1) < 0)
+        {
+            close(fd);
+            exit(99);
+        }
+        else
+            execv(command[0], command);
+        exit(99);
+    }
+    else
+    {
+        int status;
+        waitpid(pid, &status, 0);
+        if (WIFEXITED(status) && (WEXITSTATUS(status) == 99) | (WEXITSTATUS(status) == 1))
+            return false;
+
+        else
+            return true;
```

Comment:
@neohere97 think about how you could reduce duplication between the two `do_exec` implementations.